lib/private/Log/Graylog.php

```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ *
+ * @copyright Copyright (c) <2019>, <Thomas Pulzer> (t.pulzer@thesecretgamer.de)
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OC\Log;
+
+
+use OCP\IConfig;
+use OCP\ILogger;
+use OCP\Log\IWriter;
+
+class Graylog implements IWriter {
+
+	private static $VERSION = '1.1';
+	private $host;
+	private $target;
+	private $port;
+	private $protocol;
+
+	protected static $LEVELS = [
+		ILogger::FATAL => LOG_CRIT,
+		ILogger::ERROR => LOG_ERR,
+		ILogger::WARN => LOG_WARNING,
+		ILogger::INFO => LOG_INFO,
+		ILogger::DEBUG => LOG_DEBUG
+	];
+
+	public function __construct(IConfig $config) {
+		$this->host = gethostname();
+		$this->protocol = $config->getSystemValue('graylog_proto', 'udp');
+		$address = $config->getSystemValue('graylog_host', '');
+		if (false !== strpos($address, ':')) {
+			$this->target = explode(':', $address)[0];
+			$this->port = intval(explode(':', $address)[1]);
+		} else {
+			$this->target = $address;
+			$this->port = 514;
+		}
+	}
+
+	/**
+	 * send a message to the Graylog server
+	 *
+	 * @param string $app
+	 * @param string $message
+	 * @param int $level
+	 */
+	public function write(string $app, $message, int $level) {
+		$chunks = [];
+		$msg = json_encode([
+			'version' => self::$VERSION,
+			'host' => $this->host,
+			'short_message' => '{'.$app.'} '.$message,
```

Comment:
You might want to use the new way here as well:

```suggestion
			'short_message' => $this->logDetailsAsJSON($app, $message, $level),
```

Was added just now via #16416

Reply:
maybe the non-encoded one is better here: `logDetails()`, because it does not return a string with JSON, but an object with keys.

Reply:
cc @juliushaertl

Reply:
Yes, if Graylog can handle nested data in the short_message key that should be the way to go.